engine/access/rpc/backend/transaction_subscription_metadata.go

```diff
@@ -0,0 +1,315 @@
+		return err
+	}
+
+	if err := tm.refreshBlock(); err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return subscription.ErrBlockNotReady
+		}
+	}
+
+	if err := tm.refreshTransactionResult(ctx); err != nil {
+		return err
+	}
+
+	if err := tm.refreshStatus(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// refreshStatus updates the transaction's status based on its execution result.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) refreshStatus(ctx context.Context) error {
+	var err error
+
+	// Check, if transaction executed and transaction result already available
+	if tm.blockWithTx == nil {
+		tm.txResult.Status, err = tm.backendTransactions.DeriveUnknownTransactionStatus(tm.txReferenceBlockID)
+		if err != nil {
+			if !errors.Is(err, state.ErrUnknownSnapshotReference) {
+				irrecoverable.Throw(ctx, err)
+			}
+			return rpc.ConvertStorageError(err)
+		}
+		return nil
+	}
+
+	// When a block with the transaction is available, it is possible to receive a new transaction status while
+	// searching for the transaction result. Otherwise, it remains unchanged. So, if the old and new transaction
+	// statuses are the same, the current transaction status should be retrieved.
+	tm.txResult.Status, err = tm.backendTransactions.DeriveTransactionStatus(tm.blockWithTx.Height, tm.txResult.IsExecuted())
+	if err != nil {
+		if !errors.Is(err, state.ErrUnknownSnapshotReference) {
+			irrecoverable.Throw(ctx, err)
+		}
+		return rpc.ConvertStorageError(err)
+	}
+	return nil
+}
+
+// refreshBlock updates the block metadata if the transaction has been included in a block.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) refreshBlock() error {
+	if tm.txResult.CollectionID == flow.ZeroID || tm.blockWithTx != nil {
+		return nil
+	}
+
+	block, err := tm.blocks.ByCollectionID(tm.txResult.CollectionID)
+	if err != nil {
+		return err
+	}
+
+	tm.blockWithTx = block.Header
+	tm.txResult.BlockID = block.ID()
+	tm.txResult.BlockHeight = block.Header.Height
+
+	return nil
+}
+
+// refreshCollection updates the collection metadata if the transaction is included in a block.
+//
+// Parameters:
+//   - height: The block height at which the transaction is expected.
+//
+// Expected errors during normal operation:
+//   - `ErrTransactionNotInBlock` if the transaction is not found in the block.
+func (tm *transactionSubscriptionMetadata) refreshCollection(height uint64) error {
+	if tm.txResult.CollectionID != flow.ZeroID {
+		return nil
+	}
+
+	block, err := tm.blocks.ByHeight(height)
+	if err != nil {
+		return err
+	}
+
+	collectionID, err := tm.backendTransactions.LookupCollectionIDInBlock(block, tm.txResult.TransactionID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return subscription.ErrBlockNotReady
```

Comment:
I think this will cause a deadlock.

if the subscription is currently at block 100, and the tx was included in block 101, this method will return `subscription.ErrBlockNotReady`. This tells the streamer that data for this block is not ready yet, and that it needs to retry next cycle. 

At the next cycle, `refreshCollection` will be called again with block 100 which doesn't include the tx.

Instead, I think this method needs to do the lookup similar to `initBlockInfo()`, and just check if the data exists in the node's db. That way it will find the info regardless of the block passed in by the streamer. In the tx status stream case, I don't think we need to track using heights.